saq/queue/base.py

```diff
@@ -0,0 +1,342 @@
+"""
+Base Queue class
+"""
+
+from __future__ import annotations
+
+import asyncio
+import json
+import logging
+import typing as t
+from abc import ABC, abstractmethod
+from contextlib import asynccontextmanager
+
+from saq.job import (
+    TERMINAL_STATUSES,
+    UNSUCCESSFUL_TERMINAL_STATUSES,
+    Job,
+    Status,
+    get_default_job_key,
+)
+from saq.utils import now
+
+if t.TYPE_CHECKING:
+    from collections.abc import AsyncIterator, Iterable, Sequence
+
+    from saq.types import (
+        BeforeEnqueueType,
+        CountKind,
+        ListenCallback,
+        DumpType,
+        LoadType,
+        QueueInfo,
+        QueueStats,
+        VersionTuple,
+    )
+
+
+logger = logging.getLogger("saq")
+
+
+class JobError(Exception):
+    """
+    Basic Job error
+    """
+
+    def __init__(self, job: Job) -> None:
+        super().__init__(
+            f"Job {job.id} {job.status}\n\nThe above job failed with the following error:\n\n{job.error}"
+        )
+        self.job = job
+
+
+class Queue(ABC):
+    """An abstract base class for queues"""
+
+    @abstractmethod
+    def __init__(
+        self,
+        name: str,
+        dump: DumpType | None,
+        load: LoadType | None,
+    ) -> None:
+        self.name = name
+        self.started: int = now()
+        self.complete = 0
+        self.failed = 0
+        self.retried = 0
+        self.aborted = 0
+        self._version: VersionTuple | None = None
```

Comment:
Is this originally the redis version? Does it belong here?

Reply:
Yes, this should be the redis version. I'll move it to the redis queue.